source/simulator/helper_functions.cc

```diff
@@ -2114,39 +2120,83 @@ namespace aspect
+                  boundary_velocity_manager.get_active_boundary_velocity_names().end();
+
+                // We know the normal flow is zero across tangential and zero velocity boundaries.
+                // Save some time computing it
+                if (replace_noflow_boundary_ids == true &&
+                    (face_at_zero_velocity_boundary == true ||
+                     face_at_tangential_velocity_boundary == true))
+                  {
+                    face->set_boundary_id(face->boundary_id() + offset);
+                    continue;
+                  }
+
                 fe_face_values.reinit (cell, face_number);
                 fe_face_values[introspection.extractors.velocities].get_function_values(current_linearization_point,
                                                                                         face_current_velocity_values);
 
-                // ... check if the face is an outflow boundary by integrating the normal velocities
-                // (flux through the boundary) as: int u*n ds = Sum_q u(x_q)*n(x_q) JxW(x_q)...
-                double integrated_flow = 0;
+                // ... compute the the normal velocities and categorize flow across face:
+                // If any one quadrature point has inflow, the face is considered an inflow boundary,
+                // because we need to apply dirichlet boundary conditions there.
+                // On the remaining faces, if any point has outflow, the face is considered an outflow boundary.
+                // If all quadrature points have zero flow, the face is considered a noflow boundary.
+                bool is_outflow_boundary = true;
+                bool is_noflow_boundary = true;
                 for (unsigned int q=0; q<fe_face_values.n_quadrature_points; ++q)
                   {
-                    integrated_flow += (face_current_velocity_values[q] * fe_face_values.normal_vector(q)) *
-                                       fe_face_values.JxW(q);
+                    Tensor<1,dim> flow_velocity;
+                    // Use analytical boundary velocity, where known
+                    if (face_at_prescribed_velocity_boundary == true)
+                      flow_velocity = boundary_velocity_manager.boundary_velocity(face->boundary_id(), fe_face_values.quadrature_point(q));
+                    else
+                      flow_velocity = face_current_velocity_values[q];
+
+                    const double normal_flow = flow_velocity * fe_face_values.normal_vector(q);
+
+                    if (normal_flow < 0)
+                      {
+                        // inflow
+                        is_outflow_boundary = false;
+                        is_noflow_boundary = false;
+                      }
+                    else if (normal_flow > 0)
+                      {
+                        // outflow
+                        is_noflow_boundary = false;
+                      }
                   }
```

Comment:
Here, you basically look at whether there are *any* quadrature points that satisfy the condition where previously you computed an integral (=average velocity). Is that what you want? If so, why not `break` if you find one such quadrature point?